File: server/internal/handler/protocol.go
package handler

import (
	"encoding/json"
	"fmt"

	"github.com/iamkahvi/text_editor_server/internal/diff"
)

type Message struct {
	Patches   []diff.Patch    `json:"patches"`
	PatchObjs []diff.PatchObj `json:"patchObjs"`
}

func (m Message) String() string {
	payload := struct {
		Patches   []diff.Patch    `json:"patches"`
		PatchObjs []diff.PatchObj `json:"patchObjs"`
	}{
		Patches:   m.Patches,
		PatchObjs: m.PatchObjs,
	}

	b, err := json.MarshalIndent(payload, "", "    ")
	if err != nil {
		return fmt.Sprintf("Message{error: %v}", err)
	}

	return string(b)
}

// the shape of the incoming message is:
// {
//     "patches": [],
//     "patchObjs": []
// }

func (m *Message) UnmarshalJSON(data []byte) error {
	type alias struct {
		Patches   []diff.Patch    `json:"patches"`
		PatchObjs []diff.PatchObj `json:"patchObjs"`
	}

	var payload alias
	if err := json.Unmarshal(data, &payload); err != nil {
		return fmt.Errorf("unmarshal message: %w", err)
	}

	if payload.Patches == nil {
		return fmt.Errorf("message must contain patches field")
	}

	if payload.PatchObjs == nil {
		return fmt.Errorf("message must contain patchObjs field")
	}

	m.Patches = payload.Patches
	m.PatchObjs = payload.PatchObjs

	return nil
}

type MyResponse struct {
	Status string `json:"status"`
	Doc    string `json:"doc,omitempty"`
}

File: client/src/app/protocol.ts
export type Diff = [number, string];

export type PatchObj = {
  diffs: Diff[];
  start1: number | null;
  start2: number | null;
  length1: number;
  length2: number;
};

export type Patch = [-1 | 0 | 1, string];

export interface Message {
  patches: Patch[];
  patchObjs: PatchObj[];
}

export interface EditorResponse {
  type: "editor";
  status: "OK" | "ERROR";
  doc?: string;
}

export interface ClientResponse {
  type: "client";
  count: number;
}

export type MyResponse = EditorResponse | ClientResponse;
